### PR TITLE
Fix classpath with plugins java 8

### PR DESCRIPTION
**Pull request by kkrugler**
## Description
This addresses issue #7393, in a way that's hopefully compatible with both Java 8 and Java 9+. It basically reverts behavior back to how it worked previously if running with a pre-Java 9 JRE.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
No

Does this PR fix a zero-downtime upgrade introduced earlier?
No

Does this PR otherwise need attention when creating release notes? Things to consider:
No

## Release Notes
N/A

## Documentation
N/A

### CODE REVIEW

**Comment by kkrugler** [comment]
@kishoreg, @elonazoulay & @xiangfu0 - a simple attempt at fixing the problem. This works with my Hadoop segment building jobs running on Java 8, but I don't have a good environment to test against Java 9/11.

**Comment by kkrugler** [comment]
@elonazoulay & @xiangfu0 - hoping one of you can take a look, thanks!

File: pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java [hunk 1]
```diff
@@ -31,27 +32,53 @@
...
+
+    if (_classLoader instanceof URLClassLoader) {
+      try {
+        _addUrlMethod = URLClassLoader.class.getDeclaredMethod("addURL", URL.class);
+      } catch (NoSuchMethodException e) {
+        //this should never happen
+        ExceptionUtils.rethrow(e);
+      }
+      
+      _addUrlMethod.setAccessible(true);
```
Comment (richardstartin) [inline pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java:57]: Whilst this is common in the pinot codebase, doing things like this prevents use on JDK16+, creating obstacles to adoption of highly relevant JDK innovations (Vector API, foreign memory access, foreign function interface and so on).

**Comment by richardstartin** [comment]
Rather than add another illegal reflective access to the codebase, can we take a step back? Why do we need to add the URL to `PluginClassLoader`'s `ClassLoader`? Presumably to create a delegation relationship between them, so what prevents a parent-child relationship here? I do appreciate this is addressing a regression, so there will be some desire for a "quick fix" but this needs to be balanced against making it harder for pinot to run on recent JDKs.

Regardless of my future-proofing concerns, there should be a regression test which would have caught the regression in the first place, since it will cause cause some inconvenience if it regresses again.

**Comment by kkrugler** [comment]
Hi @richardstartin - yes, it's a quick fix to a regression, so I'd like to get this in for the next release. It basically restores behavior on pre-Java 9 JREs to how it was previously.

And I agree that it should be solvable in a more future-proof way. But note that the call to `_addUrlMethod.setAccessible(true)` should only happen when running with pre-Java 9 JREs.

Separately the main issue with an appropriate regression test is that this problem only shows up when running with Hadoop and Spark. Tests for those absolutely should be added, but it's non-trivial to spin up a distributed Hadoop environment for these tests (needed to test proper class loading in the context of the map test).

**Comment by richardstartin** [comment]
Regression testing this doesn't require running Hadoop: add a classfile with a class `com.Foo` in it to a location on disk, add the URL with `PluginClassloader.addURL` and then call `PluginClassloader.loadClass("com.Foo")` with the name of the test class.

**Comment by kkrugler** [comment]
@richardstartin - the "new" version of PluginClassLoader seems to work just fine when Pinot is running as a regular Java process. When we're loading plugins (via Hadoop's distributed cache) for a map task, it doesn't work. This might be due to Hadoop's classloader, which does try to do some isolation for user classes vs. Hadoop classes. It definitely warrants a deeper dive, no argument there. But after 15 years of dealing with classloader issues with Hadoop and other distributed systems, I'm just not fired up...

**Comment by richardstartin** [comment]
@kkrugler I've done this sort of thing before for bytecode instrumentation so I will try to produce a regression test without Hadoop in the picture. Probably early next week.

**Comment by kkrugler** [comment]
@richardstartin a regression test that fails with the current code would be awesome, thanks!

**Comment by kkrugler** [comment]
Bump - hoping this fix can get rolled in before the next release of Pinot. Note other users posting on Slack about what I think is the same problem.

File: pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java (hunk 1 above)
Comment (mayankshriv) [reply pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java:57]: Any alternative suggestion @richardstartin? Seems like the issue surfaced in Pinot 0.8.0 and is breaking spark ingestion job.

**Comment by richardstartin** [comment]
@kkrugler sorry for dropping this, I am working on this today.

**Comment by richardstartin** [comment]
@kkrugler I had a deeper look at this. 

I think adding the URLs to `PluginClassLoader.class.getClassLoader()` is an historical workaround for plugin classes being loaded through the wrong class loader. It's unfortunate because it breaks isolation between plugins (loading a plugin modifies `PluginClassLoader.class.getClassLoader()`'s classpath) and it's nondeterministic (when there are conflicts between different plugins, `PluginClassLoader.class.getClassLoader()` will load different classes depending on the order plugins are loaded). This mechanism also won't work on JDK9+ but it solves a problem for JDK8 users by reinstating the workaround.

I would like to rework this mechanism entirely so that plugins are isolated properly and this step isn't necessary, but don't want to get in the way of you fixing this for JDK8 users.

**Review by richardstartin — APPROVED** [review]
Not an approver but this looks ok to me as a short term fix.

**Review by mayankshriv — APPROVED** [review]

**Pull request merged**